subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -281,22 +280,15 @@ public void apply(final Project project) {
         SourceSet mainSourceSet = javaExtension.getSourceSets().create(SourceSet.MAIN_SOURCE_SET_NAME);
         projectInternal.getServices().get(ComponentRegistry.class).setMainComponent(new BuildableJavaComponentImpl(project, mainSourceSet));
 
-        BuildOutputCleanupRegistry buildOutputCleanupRegistry = projectInternal.getServices().get(BuildOutputCleanupRegistry.class);
         PublishArtifact jarArtifact = configureArchives(project, mainSourceSet);
 
         configureBuiltInTest(project, mainSourceSet);
-        configureSourceSets(javaExtension, buildOutputCleanupRegistry);
         createConsumableConfigurations(project, mainSourceSet, jarArtifact);
         configureJavaDocTask(null, mainSourceSet, project.getTasks(), javaExtension);
         configureBuild(project);
         configurePublishing(project, mainSourceSet);
     }
 
-    private static void configureSourceSets(JavaPluginExtension pluginExtension, final BuildOutputCleanupRegistry buildOutputCleanupRegistry) {
-        // Register the project's source set output directories
-        pluginExtension.getSourceSets().all(sourceSet -> buildOutputCleanupRegistry.registerOutputs(sourceSet.getOutput()));
```

Comment:
🤔 So the change here is that compilation outputs outside the build directory won't be cleaned up anymore when we detect stale files there. I don't know how many folks configure locations like this. I'd prefer if we keep cleaning up those locations when we can to avoid incorrect builds.

Reply:
I agree we should not remove this just yet for that reason. Though, this seems like a larger issue than just compilation outputs. What if other non-source-set-output tasks produce outputs outside of the build directory? Should the `BuildOutputCleanupRegistry` be more integrated into task output snapshotting? Or, to simplify things, would it be crazy to deprecate having task outputs outside of the build directory?

In general, we would like to get rid of directly registering source set outputs on the `BuildOutputCleanupRegistry`, as it can cause the compile task to be realized unnecessarily.

Reply:
Do we treat files in buildDir special anywhere else?

Reply:
Good question. We register the build directory and everything cleaned by the `clean` task in the `BuildOutputCleanupRegistry` here: https://github.com/gradle/gradle/blob/6b343a1f055814fb24c3b55bd5398a9acaf48023/subprojects/platform-base/src/main/java/org/gradle/language/base/plugins/LifecycleBasePlugin.java#L54-L63
That causes all the `clean` tasks to be realized.

The initial idea was to make this service a public API at some point, so your plugin can also add safe locations to be cleaned up.

Note that we can't allow all outputs to be cleanup up because you may have install tasks that install some binaries to, e.g. `/usr/bin`, and that directory shouldn't be cleanup up. The long-term goal may be that you'd need to opt-in as a non-well-behaved task that does something outside your build/output directories.

Note that, in general, the source set should govern the output locations and not the tasks. If somebody then messes with the task they are on their own.